Improve rule details and overall rule consistency

aemino opened the pull request:
As suggested by #736...

Though it was rather painful, I went through every single rule and tried to improve consistency and add more description in the details.  I tried to use my best judgement to infer what the preferred standards were, so let me know if something's off.

I forwent adding details to rules if I didn't feel it was necessary to elaborate.

Also, I saw some instances where a "smart apostrophe" was used; I wasn't sure whether or not to use it.

bwilkerson commented on lib/src/rules/always_put_control_body_on_new_line.dart:
```diff
@@ -6,12 +6,13 @@ import 'package:analyzer/dart/ast/ast.dart';
 From the [flutter style guide](https://flutter.io/style-guide/):
 
-**DO** Separate the control structre expression from its statement.
+**DO** separate the control structre expression from its statement.
```
While you're at it, "structre" --> "structure"

bwilkerson commented on lib/src/rules/avoid_classes_with_only_static_members.dart:
```diff
@@ -13,6 +13,10 @@ const _details = r'''
 **AVOID** defining a class that contains only static members.
 
+Creating classes with the sole purpose of providing utility or otherwise static
+methods is discouraged. Dart allows functions to be exist outside of classes
```
"be exist" --> "exist"

bwilkerson commented on lib/src/rules/avoid_return_types_on_setters.dart:
```diff
@@ -6,11 +6,14 @@ import 'package:analyzer/dart/ast/ast.dart';
-const details = r'''
 **AVOID** return types on setters.
 
+As setters do not return a value, setting the return type of one is redundant.
```
"setting" --> "declaring"

bwilkerson commented on lib/src/rules/hash_and_equals.dart:
```diff
@@ -7,12 +7,15 @@ import 'package:analyzer/dart/ast/visitor.dart';
 **DO** override `hashCode` if overriding `==`.
 
+Every object in Dart should have a `hashCode`. This is necessary so that
+every Dart object can have its `hashCode` used as a key in a map.
```
Every object has a hashCode. The reason it's important for `==` and `hashCode` to be consistent is so that a common implementation of hash maps will work correctly, and the reason to override both together is to maintain that consistency.

bwilkerson commented on lib/src/rules/iterable_contains_unrelated_type.dart:
```diff
@@ -13,10 +13,12 @@ const _desc = r'Invocation of Iterable<E>.contains with references of unrelated'
 return `false`. Strictly speaking it could evaluate to true since in Dart it
-is possible for an Iterable to contain elements of type unrelated to its
+is possible for an `Iterable` to contain elements of type unrelated to its
 parameter type, but this practice also should be avoided.
```
I believe that the last statement is not true in strong mode, and hence should be removed.

bwilkerson commented on lib/src/rules/list_remove_unrelated_type.dart:
```diff
@@ -8,15 +8,17 @@ import 'package:linter/src/analyzer.dart';
 return `false`. Strictly speaking it could evaluate to true since in Dart it
-is possible for an List to contain elements of type unrelated to its
+is possible for a `List` to contain elements of type unrelated to its
 parameter type, but this practice also should be avoided.
```
I believe that the last statement is not true in strong mode, and hence should be removed.

bwilkerson commented on lib/src/rules/omit_local_variable_types.dart:
```diff
@@ -8,12 +8,15 @@ import 'package:analyzer/dart/element/type.dart';
 import 'package:linter/src/util/dart_type_utilities.dart';
 
-const _desc = r'Omit the types for local variables.';
+const _desc = r'Consider omitting the types for local variables.';
```
While the guideline leaves room for choice, the lint rule doesn't. Do we want the description to match the guideline, or what the rule is telling users?

bwilkerson commented on lib/src/rules/prefer_asserts_in_initializer_lists.dart:
```diff
@@ -10,9 +10,10 @@ import 'package:analyzer/dart/ast/visitor.dart';
 import 'package:linter/src/analyzer.dart';
 
-const desc = 'Prefer put asserts in initializer list.';
+const _desc = r'Prefer put asserts in initializer list.';
```
"put" --> "putting"

bwilkerson commented on lib/src/rules/prefer_foreach.dart:
```diff
@@ -8,11 +8,16 @@ import 'package:analyzer/dart/element/element.dart';
 import 'package:linter/src/util/dart_type_utilities.dart';
 
-const _desc = r'Use forEach call to only apply a function to all the elements.';
+const _desc = r'Use `.forEach` to only apply a function to all the elements.';
```
I'd remove the period (both here and below).

aemino replied:
This certainly isn't the only instance where I used \`.methodName\` as a standard for referencing methods. Should I change all instances to be \`methodName\` instead?

bwilkerson replied:
That would be _my_ preference, but I won't block the merge because of it.

alexeieleusis commented on lib/src/rules/close_sinks.dart:
```diff
@@ -12,8 +12,10 @@ const _desc = r'Close instances of `dart.core.Sink`.';
-unexpected behaviors.
+**DO** invoke `close` on instances of `dart.core.Sink`.
+
+Closing instances of StreamSubscription prevents memory leaks and
```
Sink

alexeieleusis commented on lib/src/rules/avoid_types_on_closure_parameters.dart:
```diff
@@ -12,6 +12,10 @@ const _details = r'''
+Annotating types in function expressions is usually unnecessary because
+function expressions are almost always used on multiple values of the same
+type, thus making the practice redundant.
+
```
Consider adding something like: Usually the type should be inferred from the context, e.g. the function is provided as a callback argument.

aemino commented:
I went through all of the rules again and did another sweep, this time focusing on line formatting (making sure lines were cut off at 80 as efficiently as possible) and double-spaces after periods.

I also addressed all of the comments that were made; let me know if you see something else.

I'd also like to clarify -- there are some inconsistencies in the code where a special apostrophe character is used for statements like `DON'T`.  What's the preferred standard in this case?

bwilkerson commented:
I would prefer to _not_ use special characters. I suspect that they were the result of a copy-paste operation.

I'll try to take a look at the changes tomorrow (unless someone gets to it first).

kevmoo commented:
Does @aemino have commit access?

aemino commented:
@kevmoo No, I do not. I'm just a casual contributor. 🙂

kevmoo commented:
> @kevmoo No, I do not. I'm just a casual contributor. 🙂

Awesome! Just want to make sure this doesn't sit here unmerged. Thanks for your help!

bwilkerson commented on lib/src/rules/avoid_init_to_null.dart:
```diff
@@ -7,20 +7,20 @@ import 'package:analyzer/dart/ast/visitor.dart';
-gets initialized to null. This is reliably specified by the language. There's
-no concept of "uninitialized memory" in Dart. Adding `= null` is redundant and
+gets initialized to null.  This is reliably specified by the language.  There's
+no concept of "uninitialized memory" in Dart.  Adding `= null` is redundant and
```
"= null" --> "== null"

pq replied:
I agree.  I'll merge this.  @aemino (or anyone 😉 ): feel free to follow up with any fixes.  Tag and I'll be happy to review. 👍

Hixie commented:
FYI: This PR broke Flutter, because we used to refer directly to the `details` member in `package_names.dart`, which this patch makes private.
